src/instructions/mod.rs

```diff
@@ -7,184 +8,364 @@ pub mod m;
 pub mod rvc;
 
 pub use self::register::Register;
-use super::machine::Machine;
 use super::Error;
-use std::fmt::{self, Display};
+pub use execute::execute;
 
-#[derive(Debug, Clone)]
-pub enum Instruction {
-    RVC(rvc::Instruction),
-    I(i::Instruction),
-    M(m::Instruction),
+type RegisterIndex = usize;
+type Immediate = i32;
+type UImmediate = u32;
+
+// For fast decoding and cache friendly, RISC-V instruction is decoded
+// into 64 bit unsigned integer in the following format:
+//
+// +-----+-----+-----+-----+-----+-----+-----+-----+
+// |           | rs2 | rs1 | res |     | rd  | op  | R-type
+// +-----------+-----------------------------------+
+// |    immediate    | rs1 | res |     | rd  | op  | I-type
+// +-----------------------------------------------+
+// |    immediate    | rs1 | res |     | rs2 | op  | S-type/B-type
+// +-----------------+-----------------------------+
+// |       immediate       | res |     | rd  | op  | U-type/J-type
+// +-----+-----+-----+-----+-----+-----+-----+-----+
+//
+// +res+ here means reserved field that is not yet used.
+//
+// This way each op and register index are in full byte, accessing them
+// will be much faster than the original compact form. Hence we will have
+// a fast path where the interpreter loop reads instruction directly in this
+// format, and a slow path where a full featured decoder decodes RISC-V
+// instruction into the internal form here(much like how traces/micro-ops work.)
+pub type Instruction = u64;
+
+pub type InstructionOpcode = u8;
+
+// ADDI must have an opcode of 0, this way the default 0 value for u64
+// happens to be a NOP operation in RISC-V format. Of course using ADD here
+// should have the same effect as well but we just want to stick to RISC-V
+// established convention here.
+pub const OP_ADDI: InstructionOpcode = 0;
```

Comment:
Ever tried declaring an enum with `repr(u8)`?

Reply:
Just did some experiment:

* [The original unoptimized case](https://github.com/xxuejie/ckb-vm/blob/d86d1ae8e9d617ecef467d52a099713650655db8/src/instructions/mod.rs#L44): 32.242 ms
* [Adding repr(u8) alone](https://github.com/xxuejie/ckb-vm/blob/3c0675a4194d9e6150df9cbe443924c8a66114ab/src/instructions/mod.rs#L45): 32.519 ms
* [Using repr(u8) and transmute](https://github.com/xxuejie/ckb-vm/blob/c2a9303510f594490d93050806e4c37f75297a43/src/instructions/mod.rs#L46): 30.875 ms
* [Using just plain u8 constants](https://github.com/xxuejie/ckb-vm/blob/5838875739396217be73a545545105bc480ac654/src/instructions/mod.rs#L47): 27.736 ms

My guess here is that `[repr(u8)]` really just only affects the enum size here, and since we already make sure the enum is an [u8](https://github.com/xxuejie/ckb-vm/blob/d86d1ae8e9d617ecef467d52a099713650655db8/src/instructions/mod.rs#L442), it doesn't really make much difference for performance.